core/tests/test_utils.py

```diff
@@ -2454,25 +2457,50 @@ def assertRaisesRegexp(  # pylint: disable=keyword-arg-before-vararg
 class AppEngineTestBase(TestBase):
     """Base class for tests requiring App Engine services."""
 
-    # We can't instantiate the stub in setUp because the overrided
-    # method run() executes before setUp() and run() requires the memory
-    # cache stub.
-    def __init__(self, *args, **kwargs):
-        super(AppEngineTestBase, self).__init__(*args, **kwargs)
+    def setUp(self):
         self.memory_cache_services_stub = MemoryCacheServicesStub()
         self.taskqueue_services_stub = TaskqueueServicesStub(self)
 
-    def _delete_all_models(self):
-        """Deletes all models from the NDB datastore."""
-        datastore_services.delete_multi(
-            datastore_services.query_everything().iter(keys_only=True))
+        # Set up a new ExitStack. When exiting an ExitStack's context (via early
+        # return or an exception), it "unwinds" and invokes all of the callbacks
+        # and exits gathered up until that point in reverse order.
+        #
+        # NOTE: We do _NOT_ want the callbacks/exits to be invoked immediately;
+        # our goal is to do that in self.tearDown(). Regardless, we build up the
+        # stack from within its own context just in case an early exit occurs
+        # anyway (e.g. when a developer presses CTRL-C).
+        with contextlib2.ExitStack() as stack:
+            stack.enter_context(self.swap(
+                platform_taskqueue_services, 'create_http_task',
+                self.taskqueue_services_stub.create_http_task))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'flush_cache',
+                self.memory_cache_services_stub.flush_cache))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_multi',
+                self.memory_cache_services_stub.get_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'set_multi',
+                self.memory_cache_services_stub.set_multi))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'get_memory_cache_stats',
+                self.memory_cache_services_stub.get_memory_cache_stats))
+            stack.enter_context(self.swap(
+                memory_cache_services, 'delete_multi',
+                self.memory_cache_services_stub.delete_multi))
+
+            # The method stack.pop_all() clears out the contexts pushed onto the
+            # stack and returns a brand new stack that owns them instead.
+            self._stack = stack.pop_all()
+
+            # Since the old stack is empty and the new stack is not held by the
+            # current `with` context, we can exit the old stack's context while
+            # accomplishing our goal: do _NOT_ invoke the callbacks/exits yet.
+            pass # Documents end of context, so pylint: disable=unnecessary-pass
```

Comment:
You don't need this -- the comment above it is descriptive of this whole thing so you can combine it with the comment above the "with".

Reply:
Done.